modules/services/window-managers/bspwm/default.nix

```diff
@@ -61,21 +61,23 @@ in {
 
     home.packages = [ cfg.package ];
 
-    xdg.configFile."bspwm/bspwmrc".source = pkgs.writeShellScript "bspwmrc" ''
-      ${concatStringsSep "\n" (mapAttrsToList formatMonitor cfg.monitors)}
+    xdg.configFile."bspwm/bspwmrc".source = pkgs.writeShellScript "bspwmrc"
+      ((optionalString (cfg.extraConfigEarly != "") (cfg.extraConfigEarly + "\n"))
+        + ''
+          ${concatStringsSep "\n" (mapAttrsToList formatMonitor cfg.monitors)}
 
-      ${concatStringsSep "\n" (mapAttrsToList formatSetting cfg.settings)}
```

Comment:
Perhaps the lines that are currently before `${cfg.extraConfig}` could instead be added to `extraConfig` with something like `mkOrder 250` and the lines after added with `mkOrder 1750`. Then there would be no need for the `extraConfigEarly` since you could simply use something like `mkOrder 100 …` in your own configuration?

Reply:
I'm not familiar with many of the builtin functions, so correct me if I'm wrong, but would that approach only allow for a single extraConfig block in one of the possible positions? A user might want to do something custom at the start like source a colour scheme file and also do something after everything else like launch an application. If this is the case, we could always make the option an array of values whose order could be specified, but that seems excessive.

Reply:
Any thoughts on this ^? No worries if you're busy, I'm off this week though, so I have plenty of time to make changes if you'd like me to.